cli/azd/pkg/project/service_manager.go

```diff
@@ -0,0 +1,467 @@
+		var result *ServiceDeployResult
+
+		serviceEventArgs := ServiceLifecycleEventArgs{
+			Project: serviceConfig.Project,
+			Service: serviceConfig,
+		}
+
+		err := serviceConfig.Invoke(ctx, ServiceEventDeploy, serviceEventArgs, func() error {
+			restoreTask := sm.Restore(ctx, serviceConfig)
+			go syncProgress(task, restoreTask.Progress())
+			restoreResult, err := restoreTask.Await()
+			if err != nil {
+				return err
+			}
+
+			buildTask := sm.Build(ctx, serviceConfig, restoreResult)
+			go syncProgress(task, buildTask.Progress())
+			buildResult, err := buildTask.Await()
+			if err != nil {
+				return err
+			}
+
+			packageTask := sm.Package(ctx, serviceConfig, buildResult)
+			go syncProgress(task, packageTask.Progress())
+			packageResult, err := packageTask.Await()
+			if err != nil {
+				return err
+			}
+
+			publishTask := sm.Publish(ctx, serviceConfig, packageResult)
+			go syncProgress(task, publishTask.Progress())
+			publishResult, err := publishTask.Await()
+			if err != nil {
+				return err
+			}
+
+			result = &ServiceDeployResult{
+				ServicePublishResult: publishResult,
+			}
+
+			return nil
+		})
+
+		if err != nil {
+			task.SetError(fmt.Errorf("failed deploying service '%s': %w", serviceConfig.Name, err))
+			return
+		}
+
+		task.SetProgress(NewServiceProgress("Deployment completed"))
+		task.SetResult(result)
+	})
+}
+
+// GetServiceTarget constructs a ServiceTarget from the underlying service configuration
+func (sm *serviceManager) GetServiceTarget(ctx context.Context, serviceConfig *ServiceConfig) (ServiceTarget, error) {
+	var target ServiceTarget
+	if err := sm.serviceLocator.ResolveNamed(serviceConfig.Host, &target); err != nil {
+		return nil, fmt.Errorf(
+			"failed resolving service target for '%s', host '%s': %w",
+			serviceConfig.Name,
+			serviceConfig.Host,
+			err,
+		)
+	}
+
+	if err := target.Initialize(ctx, serviceConfig); err != nil {
+		return nil, fmt.Errorf("failed initializing service target for '%s': %w", serviceConfig.Name, err)
+	}
+
+	return target, nil
+}
+
+// GetFrameworkService constructs a framework service from the underlying service configuration
+func (sm *serviceManager) GetFrameworkService(ctx context.Context, serviceConfig *ServiceConfig) (FrameworkService, error) {
+	var frameworkService FrameworkService
+	if err := sm.serviceLocator.ResolveNamed(serviceConfig.Language, &frameworkService); err != nil {
+		return nil, fmt.Errorf(
+			"failed resolving framework service for '%s', language '%s': %w",
+			serviceConfig.Name,
+			serviceConfig.Language,
+			err,
+		)
+	}
+
+	// For containerized applications we use a composite framework service
+	if serviceConfig.Host == string(ContainerAppTarget) || serviceConfig.Host == string(AksTarget) {
+		var compositeFramework CompositeFrameworkService
+		if err := sm.serviceLocator.ResolveNamed(string(ServiceLanguageDocker), &compositeFramework); err != nil {
+			return nil, fmt.Errorf(
```

Comment:
I would consider `panic`ing here.  The only time this is going to fail is when someone (i.e. us, the developers of azd) have not constructed the container correctly, correct?